Add support for adding private binaries into scope

rgrinberg opened the pull request:
This PR adds support for overriding the binary in a scoped way. For example:

```
(env
 (_
  (bins (./foo/bar.exe as bar.exe))))
```

### TODO:

- [x] I just noticed that cross compilation is a bit broken currently. For example, `resolve_program` does not seem to respect it. I will fix this issue (and others like it).

- [x] `PATH` is not being modified. We've agreed that w/e binaries were added into scope using `(bins ..)` should also be added to `PATH` (without any sandboxing). This is not so simple to implement, so I might delay this to another PR.

Copying users who have shown interest in this feature: @ejgallego @samoht and @avsm

rgrinberg committed:
Move file bindings list to own module

This parser will be shared between dune_env and dune_file

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Add bins field to env stanza

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Get artifacts binary based on the env

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Tests for environment's new bin field

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Fix cross compilation issues

resolve_program and the new env based artifacts store should account for cross
compilation

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

diml commented:
Regarding `PATH`, it is indeed the case that it shouldn't be modified for individual rules, however a new directory should be added everytime we see a `(bins ...)`. Otherwise, we won't be able to use he same name in different directories, for instance:

```
foo/dune:
(env (_ (bins (foo.exe as x)))

bar/dune:
(env (_ (bins (bar.exe as x)))
```

BTW, in addition to this feature, I'd like that we also support declaring that a binary is available in the current project via the `executable` stanza directly. The reason is that in general it is not necessary in Dune to specify explicit paths, while this is not true for this feature. As a result, in your example if you rename `foo` to `foo2`, then you also have to edit the path in the `env` stanza, which is a but unfortunate.

ejgallego commented:
I guess this only works for users of `(run foo)` right?

My use case was a bit different, as I wanted to add a binary to an action, in this case the action building the sphinx docs so it could locate where coqtop is.

rgrinberg commented:
Well, this is the first step. Once I finish the 2nd todo point, the binary will in fact be in `PATH`. All that would be left would be to add support for `%{bin:..}` in deps.

diml commented:
BTW, this is touching the same code as #1354. Looking at the two features, I believe it would be easier to do the changes discussed in #1354 for the artifacts.ml module first. The general idea is that instead of manually building the list of local binaries by scanning the stanzas, we instead list the contents of the local bin directory.

rgrinberg committed:
Parameterize _ File_bindings.t and expand it in env

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
This PR is now rebased on master (which includes #1354). In the process, I've also allowed for some variable expansion in this `bins` field.

> The general idea is that instead of manually building the list of local binaries by scanning the stanzas, we instead list the contents of the local bin directory.

Do you think we should do this for the "local" (dir/env specific) binary directory as well? It seems like there's not as much benefit here as `File_bindings.t` is a one to one mapping to the contents of this directory.

rgrinberg committed:
Add dir local binary to PATH

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Add file_bindings to environment

This is necessary to setup the copy rules for the binary directory

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Add local binaries to .bin

Execute binaries from this directory as well

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Use the correct path seperator for PATH

also move to Bin to stdune. PATH is now passed explicitly where possible, as
it's too easy to forget to get the correct PATH

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
Okay, now everything works. The code is a bit of a mess right now, but it's ready for another round of review.

The last commit makes passing PATH a lot more explicit at the cost of boilerplate. In my opinion, it's worth it because it was too easy before to look up a binary without accounting for the env overrides. Now we should be able to override almost any binary in a build in a scoped way.

rgrinberg committed:
Fix 80 chars

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Add tests for locally modified PATH

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented on src/print_diff.ml:
```diff
@@ -21,7 +21,8 @@ let print ?(skip_trailing_cr=Sys.win32) path1 path2 =
   in
   let normal_diff () =
-    match Bin.which "diff" with
+    (* TODO we should be using the diff program from env *)
```
Yeah, which is why I left it as a TODO.

diml replied:
I mean, maybe it shouldn't even be a TODO. It adds more error cases to handle: for instance if the construction of the `diff` tool fails, then you end up in an awkward state. Not that it's not feasible to handle it, but I'm not sure it's worth making the code more complex than it already is.

rgrinberg committed:
Rename bins to binaries

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Add docs to local_bin and the file_bindings module

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented on src/env_node.ml:
```diff
@@ -18,8 +20,25 @@ let make ~dir ~inherit_from ~scope ~config ~env =
+  | Some x -> x
+  | None ->
+    let file_bindings =
+      match Dune_env.Stanza.find t.config ~profile with
```
Yeah, that's a good point. Will do it in a subsequent PR.

rgrinberg committed:
Do not add empty dirs to PATH for local binaries

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg committed:
Remove unnecessary .exe extension from the local artifacts map

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented on src/artifacts.ml:
```diff
@@ -49,6 +49,18 @@ let binary t ?hint ~loc name =
+        List.fold_left bindings ~init:(Lazy.force t.local_bins)
+          ~f:(fun acc fb ->
+            let path = File_bindings.dst_path fb ~dir:(Utils.local_bin dir) in
+            String.Map.add acc (Path.basename path) path))
```
I guess. But I'm a little confused about how we want this to work. We want the keys in this map to always be without the `.exe` extension, right? Then why do we only want to strip it on Windows?

I've added the .exe stripping when the `as` name isn't specified. So now the syntax is:

```
(foo.exe as bar)
```

But now I find that it's awkward that we have a target on the left and a path on the right. Do you think it's worth stripping an exe from the `as <name>` as well to have:

```
(foo.exe as bar.exe)
```

rgrinberg commented:
@diml this is ready for another round of review

rgrinberg committed:
Update manual & CHANGES regarding binaries field

Signed-off-by: Rudi Grinberg <rudi.grinberg@gmail.com>

rgrinberg commented:
It's updated.
